### DISCUSSION TITLE
How to chain states when unfolding a stream?

### DISCUSSION THREAD

**Discussion by clintonmead** [composewell/streamly#2908] [Q&A]
So I've got a function like the following:

```
getOneToken :: (ByteString, State) -> Maybe (Token, (ByteString, State))
getOneToken = _ 
```

So what this does is grabs one `Token`s slices a bit off the front of the `ByteString` (we can do this without copying with `ByteString`) and gives both a `Token`, and the remaining `ByteString`, plus some `State`.

We can then do:

```
unfoldBytestring :: Unfold m (ByteString, State) Token
unfoldBytestring = unfoldR getOneToken
```

But here's what I want to do. Lets say we have an incoming stream of bytestrings:

```
byteStringStream :: Stream m ByteString
```

What I want to do is state with an initial state say:

```
initialState :: State
initialState = _
```

and feed that into `unfoldBytestring` to get a stream of `Token`s. But once that unfold returns `Nothing`, I want to use the same unfold to unfold the next `ByteString` in `byteStringStream` into a stream of `Token`s, but this time instead seeding the unfold with `initialState`, I want to seed it with the last `State` returned from the previous fold (i.e. just before the previous unfold returned `Nothing`).

Note here the full "state" is `(ByteString, State)`, so the second unfold I want to seed with the second `ByteString` in the `unfoldBytestring` along with the last `State` from the unfolding of the first `ByteString` in `unfoldBytestring`.

Is there a function/combinator to do this? Or should I write my own? It seems close to `unfoldMany`, I just need a way to collect the final states from the unfolds and pass them along. I think the signature I need is something like this:

```
unfoldManyChain :: ((s, a) -> s) -> (a -> s) -> Unfold m s b -> Stream m a -> Stream m b
unfoldManyChain nextStateFunc initialStateFunc unfold stream
```

Where we take the first element `a` from `stream`, apply `initialStateFunc` to it to get `s`, then `unfold` that `s` to get a stream of `b`, and continue by combining the last state `s` from the previous unfold with the next `a` from `stream` to get a new state `s` which we then use `unfold` to produce more elements and so on.

**Comment by shlok** [comment]
Turning `Stream a` into a `Stream b` using a given stateful computation (a `Fold`) is something I have been doing with `scan` or `postscan`. Will this not work for you?
